magento/pwa-studio
# Improve Jest error detection and reporter output (#1615)

## Description

Improve Jest exception handling and clean up Jest reporter output.

- Remove `fileMock` from venia-concept, since it's used in `venia-ui` now.
- Add env setup file which suppresses console output from application code, _but allows console output from tests themselves_ for debugging purposes (don't leave console output in your tests please)
- Add custom environments for both jsdom and nodejs tests which capture unhandled promise rejections and fail tests if they exist
- Correct the tests that were now failing:
  - `packages/pwa-buildpack/lib/Utilities/__tests__/graphQL.spec.js` was testing promise rejections with statements like `expect(promise).rejects`. In Jest you need to add another assertion after `.rejects`.
  - `packages/venia-concept/src/__tests__/index.spec.js` was trying to re-require the module in a way that caused unhandled rejections. Refactored the implementation for testability and simplified the test.
  - `packages/venia-ui/lib/components/ProductFullDetail/__tests__/productFullDetail.spec.js` was mocking a React component in a way that made it render `undefined`, which was causing a problem in `react-test-renderer`.

### Notes:
- Unhandled promise rejections should be test failures. This solution isn't perfect, but we're trying to get there.
- If you want console output, you can either put console output in the test itself, or set the environment variables `NODE_DEBUG` to `DEBUG` to a non-empty value.
- **Please don't leave console output in your test files.** Console output in your app files themselves will be suppressed, so you don't have to worry about that.

## Related Issue
Closes #1615.

## Acceptance 
### Verification Stakeholders

All core devs and QA:
  - @jimbo  
  - @supernova-at 
  - @sirugh
  - @revanth0212 
  - @dpatil-magento 

### Verification Steps

1. Check out the branch and run a clean install.
2. Run `yarn test`.
3. Observe clean output without messy console logging.
4. Run `NODE_DEBUG=1 yarn test`.
5. Observe all the console output like before.
6. Edit a test and break it: remove an `await` from an asynchronous test to cause an unhandled rejection.
7. Run `yarn run test:dev`.
8. Observe watch mode showing broken test.
9. Correct the test and save the file.
10. Observe watch mode recovering.

Optional:

1. Add a `console.log` statement to a test file, helper file, or mock file.
2. Run `yarn test`.
3. Observe that the console message appears in reporter output even though the others are suppressed.

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.
